### PR TITLE
 gnrc_netif: add packet to queue when device is busy

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This fixes the issue showcased in #11256 (assuming the `gnrc_netif_pktq` module is compiled in and the network device supports to return `-EBUSY`) by providing a very simple MAC scheme: If the device is busy on send, queue the packet and don't send. When RX_COMPLETE or TX_COMPLETE is issued, try to send the first packet in the queue.

### Testing procedure
Run `make tests-gnrc_netif_pktq test` for `tests/unittests`. For testing the actual integration a device adaptation is needed, which I will provide for `at86rf2xx`.

Since I edited the `gnrc_netif.c`: pinging with `gnrc_networking` should still work.

### Issues/PRs references
Provides a fix for the race condition show-cased in #11256.

Basis for #11068 

Route to 6Lo minimal fragment forwarding

SY`) by providing a very simple MAC scheme: If the device is busy on send, queue the packet and don't send. When RX_COMPLETE or TX_COMPLETE is issued, try to send the first packet in the queue.

### Testing procedure
Run `make tests-gnrc_netif_pktq test` for `tests/unittests`. For testing the actual integration a device adaptation is needed, which I will provide for `at86rf2xx`.

Since I edited the `gnrc_netif.c`: pinging with `gnrc_networking` should still work.

### Issues/PRs references
Provides a fix for the race condition show-cased in #11256.

### CODE REVIEW

**Comment by bergzand** [comment]
Initial gut feeling for this API change is positive. I think it simplifies a few device drivers as those don't have to block anymore until the radio is done transmitting the previous frame. I know that at least the mrf24j40 has a clunky `while(not_in_the_right_state) {}` loop that might be refactored out with this change.

> MAC scheme

Maybe this should be called a Device Access Control Scheme? :smile:

**Comment by bergzand** [comment]
Initial gut feeling for this API change is positive. I think it simplifies a few device drivers as those don't have to block anymore until the radio is done transmitting the previous frame. I know that at least the mrf24j40 has a clunky `while(not_in_the_right_state) {}` loop that might be refactored out with this change.

> MAC scheme

Maybe this should be called a Device Access Control Scheme? :smile:

**Comment by miri64** [comment]
See #11263 for my `at86rf2xx` adaption.

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (kaspar030) [inline sys/net/gnrc/netif/gnrc_netif.c]: This seems problematic to me.
For one, it is completely surprising to see this here, without the context of why this PR exists in the first place.

The other problem would be that this basically requires the network device to keep the just received packet "on hold", send a queued packet, and only then get the held packet out of the device. I don't think all devices allow that.

I'm not sure, this might be solved by getting the received packet first, then initiating the send. E.g., swap 1444 and 1445. A downside would be that both packets need to be in memory at the same time, but I suspect that otherwise, the device needs to provide extra buffering...

**Review by kaspar030 — DISMISSED** [review]

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: > The other problem would be that this basically requires the network device to keep the just received packet "on hold", send a queued packet, and only then get the held packet out of the device. I don't think all devices allow that.

Oh damn, you are right, this needs to be called after the receive.

> I'm not sure, this might be solved by getting the received packet first, then initiating the send. E.g., swap 1444 and 1445. A downside would be that both packets need to be in memory at the same time, but I suspect that otherwise, the device needs to provide extra buffering...

Yepp. Rather put it in the stack's buffer.

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: > For one, it is completely surprising to see this here, without the context of why this PR exists in the first place.

How can I alleviate this surprise? Would some comment help?

Comment (kaspar030) [reply sys/net/gnrc/netif/gnrc_netif.c]: I guess so. IMO we need some documentation to point to that explains why the netif thread cannot block when it gets a send request (or why it cannot block at all actually).

For now a comment is probably sufficient. Something that refers to where the packets get queued.

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Done.

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (kaspar030) [inline sys/net/gnrc/netif/gnrc_netif.c]: wording nitpick here and below: only one packet is sent (not packet*s*).

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Fixed

**Comment by kaspar030** [comment]
> Initial gut feeling for this API change is positive. I think it simplifies a few device drivers as those don't have to block anymore until the radio is done transmitting the previous frame. I know that at least the mrf24j40 has a clunky `while(not_in_the_right_state) {}` loop that might be refactored out with this change.

My feelings exactly. Now we can finally make the drivers just return "-EBUSY" when it currently cannot send.

Conceptionally this is a rather large change, so I suggest we don't rush it into the release.

**Comment by miri64** [comment]
> Conceptionally this is a rather large change, so I suggest we don't rush it into the release.

Yepp, fully agree with that. I just need it for my current work in #11068 but I believe #11068 itself isn't ready for this release as well ;-).

**Comment by miri64** [comment]
Also I just noticed that it is not that easy `netif->ops->send()` currently is released no matter what so it is lost and can't be queued. This means we need to change *all* the link-layer glue-code and adapt the documentation (i.e. API change) to make this possible.

**Comment by miri64** [comment]
Rebased to current master to resolve conflict.

**Comment by miri64** [comment]
And set to WIP due to https://github.com/RIOT-OS/RIOT/pull/11263#issuecomment-476754981

**Comment by miri64** [comment]
Made queue larger, since it was way to small for larger fragmented packets.

**Comment by miri64** [comment]
Reworked a little bit, so a rewrite of all the `gnrc_netif_*` modules is not required for when `gnrc_netif_pktq` is included.

**Comment by miri64** [comment]
Rebased to current master to resolve a conflict caused by the merging of #11837.

**Comment by benpicco** [comment]
> Also I just noticed that it is not that easy `netif->ops->send()` currently is released no matter what so it is lost and can't be queued. This means we need to change all the link-layer glue-code and adapt the documentation (i.e. API change) to make this possible.

What do you mean by that? Is that within GNRC? I don't see device drivers releasing packets.

**Comment by miri64** [comment]
> > Also I just noticed that it is not that easy `netif->ops->send()` currently is released no matter what so it is lost and can't be queued. This means we need to change all the link-layer glue-code and adapt the documentation (i.e. API change) to make this possible.
> 
> What do you mean by that? Is that within GNRC? I don't see device drivers releasing packets.

Irrelevant due to the change later added in [lines 1314-1316](https://github.com/RIOT-OS/RIOT/pull/11263/files#diff-3eb20543397a7e6bb7064e71fac0471bR1314).

**Comment by miri64** [comment]
BTW the race condition I am trying to fix here might have been also fixed in https://github.com/RIOT-OS/RIOT/pull/12069 I still need to confirm that but if so, I will abandon this PR.

**Comment by miri64** [comment]
> > What do you mean by that? Is that within GNRC? I don't see device drivers releasing packets.
> 
> Irrelevant due to the change later added in [lines 1314-1316](https://github.com/RIOT-OS/RIOT/pull/11263/files#diff-3eb20543397a7e6bb7064e71fac0471bR1314).

And yes, the GNRC glue-code was meant.

**Comment by benpicco** [comment]
> BTW the race condition I am trying to fix here might have been also fixed in #12069 I still need to confirm that but if so, I will abandon this PR.

Huh? This would still prevent a packet being silently dropped if the device is busy, independent of any race condition.

**Comment by miri64** [comment]
Rebased to current master

**Comment by benpicco** [comment]
Is this still WIP?

**Comment by miri64** [comment]
Yepp, nothing changed so far.

**Comment by benpicco** [comment]
I just tried it together with my `at86rf215` driver and it worked out of the box.
Either my 120 MHz CPU is too fast or my driver is too slow, but when 6lo fragments a packet `send()` is called while the device still transmits the first fragment. So without this, the second fragment is always lost.

With this applied, it just worked™ after enabling `gnrc_netif_pktq`.

**Comment by miri64** [comment]
> I just tried it together with my `at86rf215` driver and it worked out of the box.

🤨 so your driver returns `-EBUSY` on send in any case? A concept were there isn't even a consensus on yet (which is why this is still WIP)?

**Comment by miri64** [comment]
> Either my 120 MHz CPU is too fast or my driver is too slow, but when 6lo fragments a packet `send()` is called again while the device still transmits the first fragment. So without this, the second fragment is always lost.
>
> With this applied, it just works™ when enabling `gnrc_netif_pktq`.

Why not do it like the current `at86rf2xx` implementation is doing and block until the device is not busy anymore?

**Comment by benpicco** [comment]
> > I just tried it together with my `at86rf215` driver and it worked out of the box.
> 🤨 so your driver returns `-EBUSY` on send in any case?

Yes, what else is it supposed to do?

>  A concept were there isn't even a consensus on yet (which is why this is still WIP)?

What's the alternative anyway?

**Comment by miri64** [comment]
> > > I just tried it together with my `at86rf215` driver and it worked out of the box.
> > > 🤨 so your driver returns `-EBUSY` on send in any case?
> 
> Yes, what else is it supposed to do?
> 
> > A concept were there isn't even a consensus on yet (which is why this is still WIP)?
> 
> What's the alternative anyway?

See https://github.com/RIOT-OS/RIOT/pull/11263#issuecomment-554468045

**Comment by miri64** [comment]
My thinking was to maybe make returning `-EBUSY` configurable and block if not enabled. What do you think @jia200x?

**Comment by benpicco** [comment]
> Why not do it like the current `at86rf2xx` implementation is doing and block until the device is not busy anymore?

I don't see it blocking anywhere, not in `_send()`, not in `at86rf2xx_tx_prepare()` - it seems like it just hopes for the best.

**Comment by benpicco** [comment]
Ah `at86rf2xx_set_state()` will loop.

**Comment by benpicco** [comment]
```C
#ifdef MODULE_GNRC_NETIF_PKTQ
    return -EBUSY;
#else
   /* loop */
#endif 
```

would be the solution for the driver then.

**Comment by miri64** [comment]
> ```c
> #ifdef MODULE_GNRC_NETIF_PKTQ
>     return -EBUSY;
> #else
>    /* loop */
> #endif 
> ```
> 
> would be the solution for the driver then.

Hell no!

There should be no GNRC-dependency in a driver. It's worse enough the `gnrc_sixlowpan` dependency wasn't sufficiently abstracted.

I'm thinking more of something like

```c
if (_busy(dev) {
    if (_is_blocking(dev)) {
        /* loop */
    }
    else {
        return -EBUSY;
    }
}
```

Where `_is_blocking(dev)` could be a static check (e.g. if a pseudo-module `at86rf215_blocking_send` is compiled in) or a dynamic check, like checking if a certain flag is set in the `dev` descriptor or maybe even both to make the code compile-time optimised.

**Comment by jia200x** [comment]
> My thinking was to maybe make returning -EBUSY configurable and block if not enabled [so the current behavior of most devices]. What do you think @jia200x?

I think in the long term we should aim to return -EBUSY when the device is busy instead of blocking until sending the next packet. The access to the radio is controlled by the MAC layer, thus it needs to know if it was not possible to access the device.
Also, it gives more flexibility to the upper layers: it's not the same to receive an -EBUSY with our current gnrc_netif_ieee802154 implementation and with TSCH, where the access to the radio is scheduled.

**Comment by benpicco** [comment]
Btw.: I remember now why I didn't use blocking `_send()` in `at86rf215`: Both `_send()` and `_isr()` are executed in the radio thread.

The at86rf215 doesn't have automatic retransmissions, but since there is a separate RX and TX buffer they are easy to implement in software, just set a timer to generate a `NETDEV_MSG_TYPE_EVENT` to wake up the radio thread to run `_isr()`.
Since the packet is still in the TX buffer it then just has to issue the TX command again or start the energy detect sequence if CSMA is used.

Now that means that with `ACK_REQ` the driver is still in the 'TX' state until the ACK has been received or the last retry is exhausted. Blocking the radio thread is not an option.

But to comply with the current API, I can just ignore the `ACK_REQ` if a second datagram is sent in rapid succession: I can block until the hardware is done TXing, but then stop the ACK timeout timer and consider the TX done, so I don't have to drop the next packet.

**Comment by benpicco** [comment]
It's [not pretty](https://github.com/RIOT-OS/RIOT/pull/12128/commits/723c33c0283f2238b4f9ac187055abca8e2437f9) but it does the job.

I'm still wondering what it takes for consensus that a packet queue is a sensible thing to have. 
I didn't see any objections, both @kaspar030 and @bergzand agreed that it's a good idea and I'd like to add my +1.

**Comment by miri64** [comment]
IIRC @PeterKietzmann had some concerns.

**Comment by jia200x** [comment]
I think it won't hurt to have a queue, but I would like to know how to handle MAC layers that already have queues (IEEE802.15.4, TSCH, etc). Is it OK to have duplicated queues?

**Comment by miri64** [comment]
> I think it won't hurt to have a queue, but I would like to know how to handle MAC layers that already have queues (IEEE802.15.4, TSCH, etc). Is it OK to have duplicated queues?

`gnrc_netif_pktq` is still an optional module...

**Comment by miri64** [comment]
Rebased to current master to resolve conflict caused by #13065.

**Comment by benpicco** [comment]
> IIRC @PeterKietzmann had some concerns.

I didn't see his comment anywhere on this PR.

**Comment by miri64** [comment]
Offline. However, I still plan to test if https://github.com/RIOT-OS/RIOT/pull/12728 maybe already fixes the issue and makes this PR unnecessary.

**Comment by benpicco** [comment]
But how does it make this PR unnecessary?
send does not block, so when sending two packets (or fragments) in rapid succession, the second frame *will* be dropped.

**Comment by miri64** [comment]
If #11068 works without this PR now, I rather would prefer a proper MAC layer and a more thought-through approach than this cheap hack ;-).

**Comment by bergzand** [comment]
I think that this is a good first step in the right direction. When working with both the mrf24j40 and the nrf52840 driver I had to jump through some hoops to get the send function working in blocking when the device is busy sending a previous frame. Being able to simple return -EBUSY in those cases simplifies the driver design to something that is not a mix of part async part blocking operations.

**Comment by miri64** [comment]
> I think that this is a good first step in the right direction. When working with both the mrf24j40 and the nrf52840 driver I had to jump through some hoops to get the send function working in blocking when the device is busy sending a previous frame. Being able to simple return -EBUSY in those cases simplifies the driver design to something that is not a mix of part async part blocking operations.

Yepp. Now we only need something to have this feature not activated for interfaces outright :-/. But maybe this will be better integrated with @jia200x's MAC rework. Rebasing for now.

**Comment by miri64** [comment]
Rebased to current master

**Comment by jia200x** [comment]
I used this one for #14787 and works like a charm. Ideally the queue should be in the MAC layer (the MAC layer need to prioritize in certain cases, so a generic upper layer queue might make it hard in some cases).

But I would say let's go with this as a mid term solution! :)

**Comment by miri64** [comment]
Ok, but then let me please clean it up first ^^"

**Comment by jia200x** [comment]
You can take the commits from https://github.com/RIOT-OS/RIOT/pull/14787 if you want :)

**Comment by miri64** [comment]
First, I squashed the fix I provided recently (found in [my latest experiments](https://github.com/5G-I3/ACM-ICN-2020-SFR))

**Comment by miri64** [comment]
Now rebased to current master.

**Comment by miri64** [comment]
> You can take the commits from #14787 if you want :)

What is the difference to the commits in this PR?

**Comment by miri64** [comment]
Just rebased to current master for now.

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (benpicco) [inline sys/net/gnrc/netif/gnrc_netif.c:1391]: Why do we need to disable IRQs here?

File: sys/include/net/gnrc/netif/conf.h
Comment (benpicco) [inline sys/include/net/gnrc/netif/conf.h:70]: This is only a fallback in case the driver doesn't send a notification for TX/RX done, right?

**Review by benpicco — COMMENTED** [review]
Code looks good, this will make driver development a lot easier!

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c]: Sorry, should have been clear that changes are still incoming. I moved this function to `gnrc_netif_pktq` itself now and there is an explaining comment.

**Comment by miri64** [comment]
Latest changes include:

- rename of `gnrc_netif_pktq_push()` to `gnrc_netif_pktq_push_back()` to make its functionality a bit clearer
- incoporation of the dequeue-timer into the `gnrc_netif_pktq` module
- usage of `IS_USED(MODULE_GNRC_NETIF_PKTQ)` and C-conditionals where applicable
- Kconfig migration
- test migration for latest changes and addition of tests for so far untested functions (where applicable without mock timer).

File: sys/include/net/gnrc/netif/conf.h
Comment (miri64) [reply sys/include/net/gnrc/netif/conf.h:70]: Or if it is missed for some reason, yes. Should I clarify that? Should I make the timer optional maybe?

**Comment by miri64** [comment]
(no functionality was changed btw)

**Comment by miri64** [comment]
> […] this will make driver development a lot easier!

But debugging harder if one forgets to include the queue ;-)

**Comment by miri64** [comment]
Made it configurable to not use the dequeue-timer.

File: sys/include/net/gnrc/netif/conf.h
Comment (benpicco) [reply sys/include/net/gnrc/netif/conf.h:70]: Yes a clarification that this should only be a fallback would be good.
Or if the hardware does not support TX done reporting (does such hardware exist?)

File: sys/net/gnrc/netif/Kconfig
Comment (leandrolanzieri) [inline sys/net/gnrc/netif/Kconfig:62]:
```suggestion
        that none of the notifications by the driver are missed!
```

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (leandrolanzieri) [inline sys/net/gnrc/netif/gnrc_netif.c:1636]: Here and bellow, just to be consistent :-)
```suggestion
#if IS_USED(MODULE_NETSTATS_L2) || IS_USED(MODULE_GNRC_NETIF_PKTQ)
```

File: sys/include/net/gnrc/netif.h
Comment (leandrolanzieri) [inline sys/include/net/gnrc/netif.h:178]:
```suggestion
#if IS_USED(MODULE_GNRC_NETIF_PKTQ) || defined(DOXYGEN)
```

**Comment by miri64** [comment]
Addressed @leandrolanzieri's comments.

**Comment by miri64** [comment]
Fixed issues pointed out by Murdock (including the missing new file ^^").

File: sys/include/net/gnrc/netif/conf.h
Comment (miri64) [reply sys/include/net/gnrc/netif/conf.h:70]: Done. See below.

**Comment by jia200x** [comment]
I'm back!

> What is the difference to the commits in this PR?

Probably nothing, just manually fixed the conflicts :)

**Comment by miri64** [comment]
So we can go ahead with the current version?

**Comment by jia200x** [comment]
yes :)

**Comment by miri64** [comment]
Should I squash then or are you still going through the changes?

**Comment by jia200x** [comment]
> Should I squash then or are you still going through the changes?

Sure, please squash

**Comment by miri64** [comment]
I just saw: also need to rebase ... again -.-

**Comment by miri64** [comment]
Squashed

**Comment by miri64** [comment]
And rebased and adapted to current master.

**Comment by jia200x** [comment]
I tested this one carefully using #14787 and notice something strange: if a radio returns `-EBUSY` instead of blocking, the RTT increases.

For instance, using this feature with `at86rf2xx`:
```
ping6 fe80::204:2519:1801:bd0e -i 170 -s 1024 -c 100
...

2020-09-03 16:04:53,168 # round-trip min/avg/max = 123.286/132.150/142.522 ms
```

If I modify the driver to return `-EBUSY`:
```c
diff --git a/drivers/at86rf2xx/at86rf2xx_netdev.c b/drivers/at86rf2xx/at86rf2xx_netdev.c
index 3c83a43947..f8ff12d0da 100644
--- a/drivers/at86rf2xx/at86rf2xx_netdev.c
+++ b/drivers/at86rf2xx/at86rf2xx_netdev.c
@@ -114,6 +114,10 @@ static int _send(netdev_t *netdev, const iolist_t *iolist)
     at86rf2xx_t *dev = (at86rf2xx_t *)netdev;
     size_t len = 0;
 
+    if (at86rf2xx_get_status(dev) == AT86RF2XX_STATE_BUSY_TX_ARET) {
+        return -EBUSY;
+    }
+
```
I get
```
ping6 fe80::204:2519:1801:bd0e -i 170 -s 1024 -c 100
...

2020-09-03 15:59:34,245 # round-trip min/avg/max = 130.729/140.632/155.966 ms
```

I consistently get extra ~8-10 ms if `-EBUSY` is returned, for 11 fragments (1024 kb ping).

Any idea on why this could happen?

**Comment by benpicco** [comment]
With `at86rf215` and 

<details><summary>this patch</summary>

```
--- a/drivers/at86rf215/at86rf215.c
+++ b/drivers/at86rf215/at86rf215.c
@@ -250,6 +250,10 @@ static void _block_while_busy(at86rf215_t *dev)
 
 static void at86rf215_block_while_busy(at86rf215_t *dev)
 {
+    if (IS_ACTIVE(GNRC_NETIF_PKTQ)) {
+        return;
+    }
+
     if (_tx_ongoing(dev)) {
         DEBUG("[at86rf215] Block while TXing\n");
         _block_while_busy(dev);
@@ -262,7 +266,12 @@ int at86rf215_tx_prepare(at86rf215_t *dev)
         return -EAGAIN;
     }
 
-    at86rf215_block_while_busy(dev);
+    if (IS_ACTIVE(GNRC_NETIF_PKTQ) && _tx_ongoing(dev)) {
+        return -EBUSY;
+    } else {
+        at86rf215_block_while_busy(dev);
+    }
+
     dev->tx_frame_len = IEEE802154_FCS_LEN;
 
     return 0;

```
</details>

I get better results:

#### master

```
2020-09-04 18:39:16,406 #  ping6 2001:db8::204:2519:1801:c8c5 -s 512 -c 10
2020-09-04 18:39:16,533 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=0 ttl=64 rssi=-54 dBm time=113.710 ms
2020-09-04 18:39:17,525 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=1 ttl=64 rssi=-57 dBm time=109.456 ms
2020-09-04 18:39:18,534 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=2 ttl=64 rssi=-56 dBm time=120.383 ms
2020-09-04 18:39:19,540 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=3 ttl=64 rssi=-62 dBm time=118.382 ms
2020-09-04 18:39:20,533 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=4 ttl=64 rssi=-62 dBm time=110.219 ms
2020-09-04 18:39:22,533 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=6 ttl=64 rssi=-65 dBm time=112.776 ms
2020-09-04 18:39:23,539 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=7 ttl=64 rssi=-61 dBm time=113.988 ms
2020-09-04 18:39:24,548 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=8 ttl=64 rssi=-61 dBm time=123.613 ms
2020-09-04 18:39:26,401 # 
2020-09-04 18:39:26,420 # --- 2001:db8::204:2519:1801:c8c5 PING statistics ---
2020-09-04 18:39:26,425 # 10 packets transmitted, 8 packets received, 20% packet loss
2020-09-04 18:39:26,428 # round-trip min/avg/max = 109.456/115.315/123.613 ms
```

#### gnrc_netif_pktq

```
2020-09-04 18:38:11,040 #  ping6 2001:db8::204:2519:1801:c8c5 -s 512 -c 10
2020-09-04 18:38:11,133 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=0 ttl=64 rssi=-57 dBm time=77.002 ms
2020-09-04 18:38:12,128 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=1 ttl=64 rssi=-55 dBm time=73.160 ms
2020-09-04 18:38:13,121 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=2 ttl=64 rssi=-56 dBm time=75.721 ms
2020-09-04 18:38:14,129 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=3 ttl=64 rssi=-56 dBm time=74.772 ms
2020-09-04 18:38:15,120 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=4 ttl=64 rssi=-55 dBm time=73.810 ms
2020-09-04 18:38:16,127 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=5 ttl=64 rssi=-56 dBm time=72.549 ms
2020-09-04 18:38:17,134 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=6 ttl=64 rssi=-56 dBm time=77.309 ms
2020-09-04 18:38:18,130 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=7 ttl=64 rssi=-57 dBm time=83.665 ms
2020-09-04 18:38:19,133 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=8 ttl=64 rssi=-58 dBm time=75.729 ms
2020-09-04 18:38:20,128 # 520 bytes from 2001:db8::204:2519:1801:c8c5: icmp_seq=9 ttl=64 rssi=-59 dBm time=76.030 ms
2020-09-04 18:38:20,129 # 
2020-09-04 18:38:20,131 # --- 2001:db8::204:2519:1801:c8c5 PING statistics ---
2020-09-04 18:38:20,142 # 10 packets transmitted, 10 packets received, 0% packet loss
2020-09-04 18:38:20,147 # round-trip min/avg/max = 72.549/75.974/83.665 ms
```

**Comment by aabadie** [comment]
@kaspar030 your concerns seem to be addressed. Can we dismiss your stale review ? (or you can just ACK :) )

**Comment by miri64** [comment]
> I tested this one carefully using #14787 and notice something strange: if a radio returns `-EBUSY` instead of blocking, the RTT increases.

Is it strange that when there is queuing involved that there might be an overhead?

**Comment by miri64** [comment]
> I consistently get extra ~8-10 ms if `-EBUSY` is returned, for 11 fragments (1024 kb ping).

If you are worried that it is _such_ a high overhead, you have to remember to divide your round-trip time by 22 (11 fragments send to and from) and you get a more accurate picture.

**Comment by jia200x** [comment]
> Is it strange that when there is queuing involved that there might be an overhead?

I think so. I would expect an overhead of memory usage, but not such an RTT overhead. Considering that the queue is there because network stack is much faster than the network device (including queue operations), I wouldn't expect such a time difference.

> If you are worried that it is such a high overhead, you have to remember to divide your round-trip time by 22 (11 fragments send to and from) and you get a more accurate picture.

I'm not worried about the higher RTT considering that this fixes several problems, but I'm just wondering why there's such a difference.

**Comment by jia200x** [comment]
(I'm not blocking this PR, I'm just trying to figure out where does this time difference come from)

File: sys/net/gnrc/netif/gnrc_netif.c
Comment (benpicco) [inline sys/net/gnrc/netif/gnrc_netif.c:1419]: What was meant here? :fearful:

Comment (miri64) [reply sys/net/gnrc/netif/gnrc_netif.c:1419]: When this was written, the netif implementations just threw away the packet (might not be the case with `netdev_new`, I don't know).

So, to prevent it not to be lost in that case, we `hold` the packet (i.e. the next `release` does not remove it from the packet buffer), to put it into the queue later in ll1431-1437. If the device was not busy, we just remove it in l1447.

Comment (benpicco) [reply sys/net/gnrc/netif/gnrc_netif.c:1419]: `netdev_new` does indeed [not release on send](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/netif/ieee802154/gnrc_netif_ieee802154.c#L401)

**Pull request merged**
